modules/partition-replicator/src/main/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManager.java

```diff
@@ -0,0 +1,425 @@
+
+                            throw ExceptionUtils.sneakyThrow(e);
+                        }
+
+                        return realAssignments;
+                    });
+        });
+    }
+
+    /**
+     * Check if the zone already has assignments in the meta storage locally.
+     * So, it means, that it is a recovery process and we should use the meta storage local assignments instead of calculation
+     * of the new ones.
+     */
+    private CompletableFuture<List<Assignments>> getOrCreateAssignments(
+            CatalogZoneDescriptor zoneDescriptor,
+            long causalityToken,
+            int catalogVersion
+    ) {
+        CompletableFuture<List<Assignments>> assignmentsFuture;
+
+        if (zonePartitionAssignmentsGetLocally(metaStorageMgr, zoneDescriptor.id(), 0, causalityToken) != null) {
+            assignmentsFuture = completedFuture(
+                    zoneAssignmentsGetLocally(metaStorageMgr, zoneDescriptor.id(), zoneDescriptor.partitions(), causalityToken));
+        } else {
+            assignmentsFuture = distributionZoneMgr.dataNodes(causalityToken, catalogVersion, zoneDescriptor.id())
+                    .thenApply(dataNodes -> AffinityUtils.calculateAssignments(
+                            dataNodes,
+                            zoneDescriptor.partitions(),
+                            zoneDescriptor.replicas()
+                    ).stream().map(Assignments::of).collect(toList()));
+
+            assignmentsFuture.thenAccept(assignmentsList -> LOG.info(
+                    "Assignments calculated from data nodes [zone={}, zoneId={}, assignments={}, revision={}]",
+                    zoneDescriptor.name(),
+                    zoneDescriptor.id(),
+                    Assignments.assignmentListToString(assignmentsList),
+                    causalityToken
+            ));
+        }
+
+        return assignmentsFuture;
+    }
+
+    @Override
+    public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
+        if (!ENABLED) {
+            return nullCompletedFuture();
+        }
+
+        return nullCompletedFuture();
+    }
+
+    /**
+     * Stops all resources associated with a given partition, like replicas and partition trackers.
+     *
+     * @param zonePartitionId Partition ID.
+     * @return Future that will be completed after all resources have been closed.
+     */
+    private CompletableFuture<?> stopPartition(ReplicationGroupId zonePartitionId) {
+        CompletableFuture<Boolean> stopReplicaFuture;
+
+        try {
+            stopReplicaFuture = replicaMgr.stopReplica(zonePartitionId);
+        } catch (NodeStoppingException e) {
+            // No-op.
+            stopReplicaFuture = falseCompletedFuture();
+        }
+
+        return stopReplicaFuture;
+    }
+
+    /**
+     * Stops resources that are related to provided zones.
+     *
+     * @param zones Zones to stop.
+     */
+    private void cleanUpZoneResources(Set<ReplicationGroupId> zones) {
```

Comment:
It's not zones, it's zone's partitions. I'd rather use partitionIds here. Please also fix the javadoc.